### PR TITLE
Object representation improvements 

### PR DESCRIPTION
**Pull request by marinagrprat**
**Problem**

When we call the object pyxnat it returns just a collection object and an ID number.

**Feature**

We want to return more information (metadata information) from the object. Some metadata information examples are the age, gender, handedness, insert data, insert user, number of subjects, experiments, scans...etc.

**Solution**

The changes has been done in the resources.py file.
A repr method has been added to each class (Project, Subject, Experiment, Scan and Resource).
Each method is capable to return the main information for each object dynamically (will just appear the information that the object contains, the empty fields will not appear) and a link that redirects you to the XNAT page where the object is.
This method allows us to know more information about the object immediately and find out more just clicking the link.

**Unit Tests**

For each object, 3 tests cases has been added. So in total, we have 15 unit tests in a new file tests/repr_test.py.
The tests check if the object exist, if the object does not exist and if it exists that the information that returns is correct.

### CODE REVIEW

**Comment by xgrg** [comment]
A few comments:

- On second thought, multiline representations may be a problem especially using lists, which is a typical use case in `pyxnat` (eg. lists of subjects, sessions, resources). Commits 5b93b00, 9ece9a1, bafa0e6, b8a84fe, 9a60a22 intend to provide one-line-only representations for `Project`, `Subject`, `Experiment`, `Scan` and `Resource`, respectively.

- Following this, including the description of a project in the representation was potentially breaking that one-line rule: 
bc7800d provides the `Project` class with a `desc()` function to easily access a project's description instead.

- Printing long lists was leading to some nasty overhead in execution time, due to the repetition of some generic requests (eg. `interface.select('xnat:projectData').all()`, etc..). Commits 5b93b00, 9ece9a1, bafa0e6 include some trick to cut this overhead by caching the results of these generic requests as private attributes of `Interface`.

- For some mysterious reason (@jhuguetn is aware of this), one of these requests (`xnat:subjectData`) does not include  `subject_id` in the list of attributes returned by default. 9ece9a1 therefore passes an explicit list of them to ensure `subject_id` is properly collected.

- Some PEP8-related style improvements were added: lines >80 char. shortened, some variable names revised.

Thank you @marinagrprat for a nice first PR!

**Comment by marinagrprat** [comment]
Thank you for your improvements! @xgrg

File: pyxnat/core/resources.py [hunk 1]
```diff
@@ -1593,6 +1817,27 @@ def params(self):
...
 
+    def __repr__(self):
+        resource_id = uri_last(self._uri)
+
+        # Check if resource exists
+
+        if self.exists():
+
+            # Fetch data files
+            files_count = len(self.files().fetchall())
```
Comment (jhuguetn) [inline pyxnat/core/resources.py:1828]: Not an straight-forward task but we could perhaps also consider including other attributes as well, e.g. `file_size`. See an example list of available attributes per resource object:
```
{'xnat_abstractresource_id': '14130',
 'label': 'SPM12_SEGMENT_T2T1_COREG7',
 'element_name': 'xnat:resourceCatalog',
 'category': 'resources',
 'cat_id': 'XNAT_E00001',
 'cat_desc': ' ',
 'file_count': '17',
 'file_size': '334253120',
 'tags': '',
 'content': '',
 'format': ''}
```

EDIT: I also noticed that this command below takes a decent amount of time to execute, specially for resources with a large number of files. That might be an additional argument for going the way suggested above.
```
# Fetch data files
files_count = len(self.files().fetchall())
```

File: pyxnat/tests/repr_tests.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+fp = op.join(op.dirname(op.abspath(__file__)), 'central.cfg')
+print(fp)
+central = Interface(config=fp)
+
+proj_1 = central.select.project('surfmask_smpl')
+subj_1 = proj_1.subject('CENTRAL_S01791')
+exp_1 = subj_1.experiment('CENTRAL_E04850')
+scan_1 = exp_1.scan('11')
+resource_1 = exp_1.resource('obscure_algorithm_output')
```
Comment (jhuguetn) [inline pyxnat/tests/repr_tests.py:15]: Perhaps might be a bit risky using external existing resources in Central XNAT for the CI tests, what if the owner modifies/removes them (I doubt so though)? What you think?

**Review by jhuguetn — COMMENTED** [review]
Looks very good overall, nice job here @marinagrprat and @xgrg. Find few suggestions and comments from my side in this review.

File: pyxnat/core/resources.py [hunk 3]
```diff
@@ -1388,13 +1445,100 @@ def aliases(self):
...
+        """
+        intf = self._intf
+        if hasattr(intf, '_projectData'):
+            data = getattr(intf, '_projectData')
+        else:
+            data = intf.select('xnat:projectData').all().data
+            setattr(intf, '_projectData', data)
+
+        data = [e for e in data if e['id'] == self._urn][0]
+        return data['description']
```
[the remark this answers is not in the document]
Comment (xgrg) [reply pyxnat/core/resources.py:1463]: I agree that from that perspective we may not need such function. But on second thought maybe that function would just help make the information more accessible?

File: pyxnat/core/resources.py [hunk 4]
```diff
@@ -1388,13 +1445,100 @@ def aliases(self):
...
 
 @add_metaclass(ElementType)
 class Subject(EObject):
 
     def datatype(self):
         return 'xnat:subjectData'
 
+    def __repr__(self):
+        interface = self._intf
+        subject_id = uri_last(self._uri)
```
[the remark this answers is not in the document]
Comment (xgrg) [reply pyxnat/core/resources.py:1474]: Let us see how the response times are impacted.

File: pyxnat/tests/repr_tests.py (hunk 2 above)
Comment (xgrg) [reply pyxnat/tests/repr_tests.py:15]: Indeed, it also went through my mind. I think we may take the risk, or maybe @marinagrprat could run a `sessionmirror.py` on that experiment to make a backup copy of it in `nosetests`?

File: pyxnat/core/resources.py (hunk 1 above)
Comment (xgrg) [reply pyxnat/core/resources.py:1821]: Could be a good idea indeed. And what about using the `file_count` attribute?
I don't remember how I got to use `fetchall()`. I just would like to make sure that these `file_count` and `file_size` fields always exist, not like those we found a few days ago (`ses`, etc.). What do you think?

Comment (marinagrprat) [reply pyxnat/core/resources.py:1821]: I will check if both fields always exists and if that is the case we get both attributes from there.

File: pyxnat/tests/repr_tests.py (hunk 2 above)
Comment (marinagrprat) [reply pyxnat/tests/repr_tests.py:15]: sure!

File: pyxnat/core/resources.py (hunk 1 above)
Comment (jhuguetn) [reply pyxnat/core/resources.py:1821]: Great. 

Yes, @xgrg, that's my idea as well to include both, see the comment's EDIT. Basically the problem I found is that I can't get those elements with an elegant (`KeyError` exception), 
```
resource.attrs.mget(['file_count', 'file_size'])
```
The XNAT REST API command `[GET] '/data/experiments/EXP/resources&format=FORMAT'` works fine however. Seems like we found a new issue?

As a pragmatic approach, we could do something as follows:
```
def __repr__(self):
        resource_id = self.id()

        # Check if resource exists

        if self.exists():

            # Fetch data files
            base_url = uri_parent(self._uri)
            resources = self._intf._get_json(base_url)
            res_info = [r for r in resources 
                             if r['xnat_abstractresource_id'] == resource_id][0]

            # Creating the output string to be returned
            output = '<{cl} Object> {id} `{label}` ({fc} files {fs} MB)'
            output = output.format(label=self.label(),
                                   cl=self.__class__.__name__,
                                   id=resource_id,
                                   fc=res_info['file_count'],
                                   fs=float(res_info['file_size'])/1024*2)
            return output
        else:
            return '<%s Object> %s' % (self.__class__.__name__,
                                       unquote(uri_last(self._uri)))
```

File: pyxnat/core/resources.py [hunk 5]
```diff
@@ -1593,6 +1817,27 @@ def params(self):
 @add_metaclass(ElementType)
 class Resource(EObject):
 
+    def __repr__(self):
+        resource_id = uri_last(self._uri)
```
[the remark this answers is not in the document]
Comment (xgrg) [reply pyxnat/core/resources.py:1814]: `self.id()` returns an error if the object does not exist. For now, 02b8314d5f reverts the suggestion.

**Comment by xgrg** [comment]
Some details on the last commits:
- some small PEP8 style fixes
- Added filesize to `__repr__` resource (following https://github.com/pyxnat/pyxnat/pull/141#discussion_r507544402)
- Added experiment label (following https://github.com/pyxnat/pyxnat/pull/141#discussion_r506701881)
- Removed the cache workaround introduced in 5b93b00, 9ece9a1 and bafa0e6 (after it did not prove very useful)
- Call to `id()` after checking for object existence (avoids [issue](https://github.com/pyxnat/pyxnat/pull/141#discussion_r506700757) suggested by @jhuguet)
- `subject_label` in `Experiment` repr (was fetched incorrectly)
- Renamed Project's `desc()` to `description()`
- `Attrs` always fetch `id` from parent object (avoid issues w/ deleted objects)
- Renamed `nosetests3` to `nosetests5` (because some smart one deleted project `nosetests3` on CENTRAL... :facepalm: )
- When subject level missing, add it instead of going for the shorter URL (avoids [issue](https://github.com/pyxnat/pyxnat/issues/142) mentioned by @jhuguet)

**Comment by marinagrprat** [comment]
looks great! thanks @xgrg

File: pyxnat/tests/repr_tests.py (hunk 2 above)
Comment (jhuguetn) [reply pyxnat/tests/repr_tests.py:15]: Session cloned into `nosetests5`, see `CENTRAL05_E02680`

Comment (jhuguetn) [reply pyxnat/tests/repr_tests.py:15]: Session cloned into `nosetests5` (see `CENTRAL05_E02680`). Solved.

**Review by jhuguetn — CHANGES_REQUESTED** [review]
All good. I just included one minor suggestion and a couple of additional changes,
* 468046d: Improved mechanism for collecting Scan details (~3x faster)
* cc3b425: Outer resources from Central which are used in CI tests were mirrored into a privately owned project. Testing details updated.

File: pyxnat/tests/resources_test.py [hunk 6]
```diff
@@ -6,6 +6,7 @@
 from pyxnat import Interface
 from . import skip_if_no_network
 from nose import SkipTest
+import time
```
[the remark this answers is not in the document]
Comment (xgrg) [reply pyxnat/tests/resources_test.py:9]: This one may be safely removed. My bad. Thanks.

**Review by jhuguetn — APPROVED** [review]

**Pull request merged**